wrappers/golang_v3/core/msm.go

```diff
@@ -0,0 +1,134 @@
+	///
+	PrecomputeFactor int32
+
+	/// `c` value, or "window bitsize" which is the main parameter of the "bucket method"
+	/// that we use to solve the MSM problem. As a rule of thumb, larger value means more on-line memory
+	/// footprint but also more parallelism and less computational complexity (up to a certain point).
+	/// Currently pre-computation is independent of `c`, however in the future value of `c` here and the one passed into the
+	/// `precompute_bases` function will need to be identical. Default value: 0 (the optimal value of `c` is chosen automatically).
+	C int32
+
+	/// Number of bits of the largest scalar. Typically equals the bitsize of scalar field, but if a different
+	/// (better) upper bound is known, it should be reflected in this variable. Default value: 0 (set to the bitsize of scalar field).
+	Bitsize int32
+
+	BatchSize                int32
+	AreBasesShared           bool
+	areScalarsOnDevice       bool
+	AreScalarsMontgomeryForm bool
+	areBasesOnDevice         bool
+	AreBasesMontgomeryForm   bool
+	areResultsOnDevice       bool
+
+	/// Whether to run the MSM asynchronously. If set to `true`, the MSM function will be non-blocking
+	/// and you'd need to synchronize it explicitly by running `cudaStreamSynchronize` or `cudaDeviceSynchronize`.
+	/// If set to `false`, the MSM function will block the current CPU thread.
+	IsAsync bool
+	Ext     runtime.ConfigExtensionHandler
+}
+
+func GetDefaultMSMConfig() MSMConfig {
+	return MSMConfig{
+		StreamHandle:             nil,
+		PrecomputeFactor:         1,
+		C:                        0,
+		Bitsize:                  0,
+		BatchSize:                1,
+		AreBasesShared:           true,
+		areScalarsOnDevice:       false,
+		AreScalarsMontgomeryForm: false,
+		areBasesOnDevice:         false,
+		AreBasesMontgomeryForm:   false,
+		areResultsOnDevice:       false,
+		IsAsync:                  false,
+		Ext:                      nil,
+	}
+}
+
+func MsmCheck(scalars HostOrDeviceSlice, bases HostOrDeviceSlice, cfg *MSMConfig, results HostOrDeviceSlice) (unsafe.Pointer, unsafe.Pointer, unsafe.Pointer, int) {
+	if bases.Len()%int(cfg.PrecomputeFactor) != 0 {
+		errorString := fmt.Sprintf(
+			"Precompute factor %d does not divide the number of bases %d",
+			cfg.PrecomputeFactor,
+			bases.Len(),
+		)
+		panic(errorString)
+	}
+	scalarsLength, basesLength, resultsLength := scalars.Len(), bases.Len()/int(cfg.PrecomputeFactor), results.Len()
+	if scalarsLength%basesLength != 0 {
+		errorString := fmt.Sprintf(
+			"Number of bases %d does not divide the number of scalars %d",
+			basesLength,
+			scalarsLength,
+		)
+		panic(errorString)
+	}
+	if scalarsLength%resultsLength != 0 {
+		errorString := fmt.Sprintf(
+			"Number of results %d does not divide the number of scalars %d",
+			resultsLength,
+			scalarsLength,
+		)
+		panic(errorString)
+	}
+
+	// cfg.basesSize = int32(basesLength)
+	cfg.AreBasesShared = basesLength < scalarsLength
+	cfg.BatchSize = int32(resultsLength)
```

Comment:
It is needed in `msm.PrecomputeBases` since it doesn't know about scalar's length. There can be another approach: send scalars with points in `msm.PrecomputeBases` function inside bindings to compute cfg